neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py

```diff
@@ -1,250 +1,331 @@
-            engine='python'
+        sentences : list[list[str]]
+            List with the train, validation and test sentences.
+        labels : list[list[int]]
+            List with the train, validation and test labels.
+        n_classes : int
+            Total number of classes in our datasets.
+        n_qubits_noun : int
+            Number of qubits per noun type.
+        n_qubits_sentence : int
+            Number of qubits per sentence type.
+        n_layers : int
+            Number of layers of the circuits.
+        n_single_qubit_params : int
+            Number of rotations per single qubit.
+        optimiser : torch.optim.Optimizer
+            Optimiser to use for training the model.
+        epochs : int
+            Number of epochs to be done in training. 
+        batch_size : int
+            Batch size to use in training.
+        loss_function : Callable
+            Loss function to use in training. If None,
+            a class default method (cross_entropy_loss_wrapper)
+            defined in the class will be used.
+        optimiser_args : dict
+            Optional arguments for the optmiser. 
+        device : int
+            CUDA device ID used for tensor operation speed-up.
+            -1 will use the CPU.
+        seed : int
+            Random seed for generating the initial parameters. 
+        """    
+        if n_classes > 2 ** n_qubits_sentence:
+            raise ValueError(
+                "Base 2 logarithm of the number of sentence qubits"\
+                "must be greater than the number of classes.")
+        self.sentences_train = sentences[0]
+        self.sentences_val = sentences[1]
+        self.sentences_test = sentences[2]
+        self.labels_train = labels[0]
+        self.labels_val = labels[1]
+        self.labels_test = labels[2]
+        self.n_classes = n_classes
+        self.n_qubits_sentence = n_qubits_sentence
+        self.ansatz = self.ansatz_functions_dict[ansatz](
+            {
+                lambeq.AtomicType.NOUN : n_qubits_noun,
+                lambeq.AtomicType.SENTENCE : n_qubits_sentence
+            },
+            n_layers = n_layers,
+            n_single_qubit_params = n_single_qubits_params
         )
-        labels = []
-        sentences = df['sentence'].tolist()
-
-        for i in range(df.shape[0]):
-            if df['label'].iloc[i] == 1:
-                labels.append([1,0])
-            else:
-                labels.append([0,1])
+        self.optimiser = self.optimisers_functions_dict[optimiser]
+        self.epochs = epochs
+        self.batch_size = batch_size
+        if loss_function == None:
+            self.loss_function = self.cross_entropy_loss_wrapper()
+        else:
+            self.loss_function = loss_function
+        self.optimiser_args = optimiser_args
+        self.device = device
+        self.seed = seed
 
-        return sentences, labels
+    def cross_entropy_loss_wrapper(self)-> None:
```

Comment:
Do you think you could unwrap this? Nested functions aren't a great thing.

Reply:
No afaik. Lambeq requires the loss function to have two inputs (y_hat,y). For multiclass classification, the loss function will depend on the number of classes, which is an attribure of the class (see reshape_output_lambeq_model in line 130). That's why I've used the wrapper, so that loss function takes self as well as argument.

Of course if you know another way of doing this let me know. this was the best I could think of.